### PR TITLE
[IMP] (pos_,sale_)loyalty: improve validity period

### PR DESCRIPTION
**Pull request by vchu-odoo**
Currently it is only possible to define the end_date that does not allow to prepare the program in advance.
Added the start_date property to the program model.

task-3305712

### CODE REVIEW

File: addons/pos_loyalty/models/pos_config.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ def use_coupon_code(self, code, creation_date, partner_id):
         check_date = fields.Date.from_string(creation_date[:11])
         if (coupon.expiration_date and coupon.expiration_date < check_date) or\
             (coupon.program_id.date_to and coupon.program_id.date_to < fields.Date.context_today(self)) or\
+            (coupon.program_id.date_from and coupon.program_id.date_from > fields.Date.context_today(self)) or\
```
Comment (Demesmaeker) [inline addons/pos_loyalty/models/pos_config.py:92]: Instead of checking every time if we are after the starting date if there is one and before the ending time if there is one, it might be interesting to write a method for that instead, it'll be cleaner/more readable IMO.

File: addons/loyalty/models/loyalty_program.py [hunk 2]
```diff
@@ -93,6 +94,11 @@ class LoyaltyProgram(models.Model):
             'Max usage must be strictly positive if a limit is used.'),
     ]
 
+    @api.constrains('date_from', 'date_to')
+    def _check_date_from_date_to(self):
+        if any(program.date_to and program.date_from and program.date_from > program.date_to for program in self):
+            raise UserError(_("The Start Date of the Validity Period must be anterior to the End Date."))
```
Comment (Demesmaeker) [inline addons/loyalty/models/loyalty_program.py:100]: 100 chars max =]

File: addons/loyalty/models/loyalty_program.py [hunk 3]
```diff
@@ -50,7 +50,8 @@ class LoyaltyProgram(models.Model):
         ('next_order_coupons', 'Next Order Coupons')],
         default='promotion', required=True,
     )
-    date_to = fields.Date(string='Validity')
+    date_from = fields.Date(string='Start Date')
+    date_to = fields.Date(string='End date')
```
Comment (Demesmaeker) [inline addons/loyalty/models/loyalty_program.py:54]:
```suggestion
    date_from = fields.Date(string="Start Date")
    date_to = fields.Date(string="End date")
```
By convention, strings that can be read by the users are between `"`, others by `'`

File: addons/sale_loyalty/models/sale_order.py [hunk 4]
```diff
@@ -989,7 +991,8 @@ def _try_apply_code(self, code):
                 return {'error': _('This coupon has already been used.')}
             program = coupon.program_id
 
-        if not program or not program.active:
+        if not program or not program.active or\
+            (program.date_from and program.date_from > fields.Date.context_today(self)):
```
Comment (Demesmaeker) [inline addons/sale_loyalty/models/sale_order.py:995]: Maybe add this as a 3rd case, when the code is valid, but not yet active?

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 5]
```diff
@@ -233,8 +233,19 @@ def test_program_rules_validity_dates_and_uses(self):
         self._auto_rewards(order, self.immediate_promotion_program)
         self.assertEqual(len(order.order_line.ids), 2, "The promo offer shouldn't have been applied we're not between the validity dates")
 
+        self.immediate_promotion_program.write({
+            'date_to': False,
+            'date_from': date.today() + timedelta(days=2),
+            'limit_usage': True,
+            'max_usage': 1,
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:240]: These lines are the same as the first created ones. They are not necessary when you update a record.

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

Regarding the test, it would make sense to separate the test in 2:
- test the number of uses
- test the date validity (before/between/after dates + when no starting/ending dates)

File: addons/pos_loyalty/static/src/js/Loyalty.js [hunk 6]
```diff
@@ -860,6 +863,9 @@ patch(Order.prototype, "pos_loyalty.Order", {
         if (program.is_nominative && !this.get_partner()) {
             return false;
         }
+        if (program.date_from && program.date_from >= new Date()) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/pos_loyalty/static/src/js/Loyalty.js:868]: Maybe it would make sense to add help only for the last day, because for the start date it is expected that the program is valid on start date.

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/pos_loyalty/static/src/js/Loyalty.js:868]: We should never expect the users to know what's expected and what's not, because that's a cultural thing. 1st floor in Japan is the ground floor, and that's one of the best example I got. :wink:

File: addons/loyalty/models/loyalty_program.py [hunk 7]
```diff
@@ -50,7 +50,8 @@ class LoyaltyProgram(models.Model):
         ('next_order_coupons', 'Next Order Coupons')],
         default='promotion', required=True,
     )
-    date_to = fields.Date(string='Validity')
+    date_from = fields.Date(string="Start Date", help="The program is active including start date")
+    date_to = fields.Date(string="End date", help="The program is active including end date")
```
Comment (Demesmaeker) [inline addons/loyalty/models/loyalty_program.py:54]: Could you use another word than "active" here? It's a bit confusing IMO
Or maybe something more like: "The start date is included in the validity period of this program"

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 8]
```diff
@@ -206,13 +206,11 @@ def test_program_rules_promotion_use_best(self):
         self.assertEqual(len(discounts), 1, "The order should contains the Product A line and a discount")
         self.assertTrue('Discount: 10% on your order' in discounts.pop(), "The discount should be a 10% discount")
 
-    def test_program_rules_validity_dates_and_uses(self):
-        # Test case: Based on the validity dates and the number of allowed uses
-
+    def test_program_rules_validity_dates(self):
+        # Test case: Based on the validity dates
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:210]: This is now an useless comment, that's the name of the test ^^

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 9]
```diff
@@ -231,36 +229,86 @@ def test_program_rules_validity_dates_and_uses(self):
             })
         ]})
         self._auto_rewards(order, self.immediate_promotion_program)
-        self.assertEqual(len(order.order_line.ids), 2, "The promo offer shouldn't have been applied we're not between the validity dates")
+        self.assertEqual(len(order.order_line.ids), 2,
+                         "The promo shouldn't have been applied as it is expired.")
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:233]:
```suggestion
        msg = "The promo is expired and shouldn't have been applied."
        self.assertEqual(len(order.order_line.ids), 2, msg=msg)
```

or at least:

```suggestion
        self.assertEqual(
            len(order.order_line.ids), 2, "The promo shouldn't have been applied as it is expired."
        )
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 10]
```diff
@@ -206,13 +206,11 @@ def test_program_rules_promotion_use_best(self):
...
-        # Test case: Based on the validity dates and the number of allowed uses
-
+    def test_program_rules_validity_dates(self):
+        # Test case: Based on the validity dates
+        # Test date_to (no date_from)
         self.immediate_promotion_program.write({
             'date_to': date.today() - timedelta(days=2),
-            'limit_usage': True,
-            'max_usage': 1,
         })
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:214]:
```py
        self.immediate_promotion_program.write({'date_to': date.today() - timedelta(days=2)})
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 11]
```diff
@@ -231,36 +229,86 @@ def test_program_rules_validity_dates_and_uses(self):
...
+        self.assertEqual(len(order.order_line.ids), 2,
+                         "The promo shouldn't have been applied as it is not active yet.")
+
+        self.immediate_promotion_program.write({
+            'date_from': date.today() - timedelta(days=2),
+        })
+
+        self._auto_rewards(order, self.immediate_promotion_program)
+        self.assertEqual(len(order.order_line.ids), 3,
+                         "The promo should have been applied we're between the validity dates.")
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:259]:
```suggestion
        self.immediate_promotion_program.write({
            'date_from': date.today() - timedelta(days=2),
        })
        self._auto_rewards(order, self.immediate_promotion_program)
        self.assertEqual(len(order.order_line.ids), 3,
                         "The promo should have been applied we're between the validity dates.")
```

Let's make a single block for every case, it's easier to read =]

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 12]
```diff
@@ -231,36 +229,86 @@ def test_program_rules_validity_dates_and_uses(self):
...
+            'date_to': False,
+        })
+
+        self._auto_rewards(order, self.immediate_promotion_program)
+        self.assertEqual(len(order.order_line.ids), 2,
+                         "The promo shouldn't have been applied as it is not active yet.")
+
+        self.immediate_promotion_program.write({
+            'date_from': date.today() - timedelta(days=2),
+        })
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:255]:
```suggestion
        self.immediate_promotion_program.write({'date_from': date.today() - timedelta(days=2)})
```

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello! :duck: 

Maybe test when the validity is between today and today too, once it's confirmed that's what's expected? :thinking: 

Also, it might be easier in the beginning to set the variables `past_day`, `today`, `future_day` at the beginning and then use it for the whole test (including in the order date to avoid improbable errors when the test runs at midnight).

File: addons/loyalty/models/loyalty_program.py [hunk 13]
```diff
@@ -50,7 +50,14 @@ class LoyaltyProgram(models.Model):
         ('next_order_coupons', 'Next Order Coupons')],
         default='promotion', required=True,
     )
-    date_to = fields.Date(string='Validity')
+    date_from = fields.Date(
+        string="Start Date",
+        help="The start date is included in the validity period of this program"
+    )
```
Comment (Demesmaeker) [inline addons/loyalty/models/loyalty_program.py:56]: Nitpicking

```suggestion
    date_from = fields.Date(
        string="Start Date",
        help="The start date is included in the validity period of this program",
    )
```
The comma here is so that if you add another line, you can put it after without having a 2 lines diff =]
And here, we are in the exception we talk about, but you can still see both, especially in loyalty since we didn't clean it yet, but this would also be accepted:

```suggestion
    date_from = fields.Date(
        string="Start Date",
        help="The start date is included in the validity period of this program")
```

File: addons/loyalty/models/loyalty_program.py [hunk 14]
```diff
@@ -93,6 +100,13 @@ class LoyaltyProgram(models.Model):
             'Max usage must be strictly positive if a limit is used.'),
     ]
 
+    @api.constrains('date_from', 'date_to')
+    def _check_date_from_date_to(self):
+        if any(p.date_to and p.date_from and p.date_from > p.date_to for p in self):
+            raise UserError(_(
+                "The validity period's start date must be anterior or equal to its end date."
+                ))
```
Comment (Demesmaeker) [inline addons/loyalty/models/loyalty_program.py:108]:
```suggestion
            ))
```

File: addons/pos_loyalty/models/pos_config.py [hunk 15]
```diff
@@ -87,22 +87,26 @@ def use_coupon_code(self, code, creation_date, partner_id):
...
+        error_message = False
         if (coupon.expiration_date and coupon.expiration_date < check_date) or\
-            (coupon.program_id.date_to and coupon.program_id.date_to < fields.Date.context_today(self)) or\
+            (coupon.program_id.date_to and coupon.program_id.date_to < today_date) or\
             (coupon.program_id.limit_usage and coupon.program_id.total_order_count >= coupon.program_id.max_usage):
+            error_message = _('This coupon is expired (%s).', code)
+        elif coupon.program_id.date_from and coupon.program_id.date_from > today_date:
+            error_message = _('This coupon is not yet valid (%s).', code)
+        elif not coupon.program_id.reward_ids or\
+                not any(reward.required_points <= coupon.points for reward in coupon.program_id.reward_ids):
```
Comment (Demesmaeker) [inline addons/pos_loyalty/models/pos_config.py:99]: Nitpicking

```suggestion
        elif not coupon.program_id.reward_ids or not any(
            reward.required_points <= coupon.points for reward in coupon.program_id.reward_ids
        ):
```
or
```suggestion
        elif not coupon.program_id.reward_ids or\
            not any(r.required_points <= coupon.points for r in coupon.program_id.reward_ids):
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 16]
```diff
@@ -231,36 +228,76 @@ def test_program_rules_validity_dates_and_uses(self):
...
+            'date_from': past_day,
+            'date_to': today - timedelta(days=1),
+        })
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo offer shouldn't have been applied as it is expired."
+        self.assertEqual(len(order.order_line.ids), 2, msg)
+
+        self.immediate_promotion_program.write({'date_from': today, 'date_to': today})
+        self._auto_rewards(order, self.immediate_promotion_program)
+        msg = "The promo should have been applied as we're between the validity dates"
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:276]:
```suggestion
        msg = "The promo should have been applied as today is a valid starting and ending date."
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 17]
```diff
@@ -231,36 +228,76 @@ def test_program_rules_validity_dates_and_uses(self):
...
                 'product_uom_qty': 1.0,
             })
         ]})
         self._auto_rewards(order, self.immediate_promotion_program)
-        self.assertEqual(len(order.order_line.ids), 3, "The promo offer should have been applied as we're between the validity dates")
-        order = self.env['sale.order'].create({'partner_id': self.env['res.partner'].create({'name': 'My Partner'}).id})
+        self.assertEqual(len(order.order_line.ids), 2, "The promo offer should have been applied")
+
+        order = self.env['sale.order'].create({
+                    'partner_id': self.env['res.partner'].create({'name': 'My Partner'}).id})
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:298]:
```suggestion
        order = self.env['sale.order'].create({
            'partner_id': self.env['res.partner'].create({'name': 'My Partner'}).id
        })
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 18]
```diff
@@ -271,4 +308,5 @@ def test_program_rules_validity_dates_and_uses(self):
         # Invalidate total_order_count
         self.immediate_promotion_program.invalidate_recordset(['order_count', 'total_order_count'])
         self._auto_rewards(order, self.immediate_promotion_program)
-        self.assertEqual(len(order.order_line.ids), 2, "The promo offer shouldn't have been applied as the number of uses is exceeded")
+        self.assertEqual(len(order.order_line.ids), 1,
+                         "The promo offer shouldn't have been applied as the number of uses is exceeded")
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_program_rules.py:312]:
```suggestion
        msg = "The promo offer shouldn't have been applied as the number of uses is exceeded"
        self.assertEqual(len(order.order_line.ids), 1, msg)
```

**Review by Demesmaeker — COMMENTED** [review]
Hello :duck: 

There are only nitpicking comments left ^^

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 19]
```diff
@@ -231,36 +228,77 @@ def test_program_rules_validity_dates_and_uses(self):
...
                 'product_uom_qty': 1.0,
             })
         ]})
         self._auto_rewards(order, self.immediate_promotion_program)
-        self.assertEqual(len(order.order_line.ids), 3, "The promo offer should have been applied as we're between the validity dates")
-        order = self.env['sale.order'].create({'partner_id': self.env['res.partner'].create({'name': 'My Partner'}).id})
+        self.assertEqual(len(order.order_line.ids), 2, "The promo offer should have been applied")
+
+        order = self.env['sale.order'].create({
+                    'partner_id': self.env['res.partner'].create({'name': 'My Partner'}).id
```
Comment (Feyensv) [inline addons/sale_loyalty/tests/test_program_rules.py:298]: over-indented ;)

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 20]
```diff
@@ -231,36 +228,77 @@ def test_program_rules_validity_dates_and_uses(self):
...
+        })
+
         order.write({'order_line': [
-            (0, False, {
-                'product_id': self.product_A.id,
-                'name': '1 Product A',
-                'product_uom': self.uom_unit.id,
-                'product_uom_qty': 10.0,
-            }),
             (0, False, {
```
Comment (Feyensv) [inline addons/sale_loyalty/tests/test_program_rules.py:302]:
```suggestion
            Command.create({
```

File: addons/sale_loyalty/tests/test_program_rules.py [hunk 21]
```diff
@@ -231,36 +228,77 @@ def test_program_rules_validity_dates_and_uses(self):
...
+
+    def test_program_rules_number_of_uses(self):
+        # Test case: Based on the number of allowed uses
+        self.immediate_promotion_program.write({
+            'limit_usage': True,
+            'max_usage': 1,
+        })
+        order = self.empty_order
         order.write({'order_line': [
             (0, False, {
```
Comment (Feyensv) [inline addons/sale_loyalty/tests/test_program_rules.py:287]:
```suggestion
            Command.create({
```
You can replace the old magic commands by the new "cleaner" API

File: addons/sale_loyalty/models/sale_order.py [hunk 22]
```diff
@@ -991,6 +993,8 @@ def _try_apply_code(self, code):
 
         if not program or not program.active:
             return {'error': _('This code is invalid (%s).', code), 'not_found': True}
+        elif (program.date_from and program.date_from > fields.Date.context_today(self)):
+            return {'error': _('This code is not yet active (%s).', code)}
         elif (program.limit_usage and program.total_order_count >= program.max_usage) or\
             (program.date_to and program.date_to < fields.Date.context_today(self)):
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:999]: `fields.Date.context_today(self)` can be extracted in a variable before the if

File: addons/sale_loyalty/models/sale_order.py [hunk 23]
```diff
@@ -392,6 +393,7 @@ def _get_trigger_domain(self):
         self.ensure_one()
         return [('active', '=', True), ('program_id.sale_ok', '=', True),
                 ('company_id', 'in', (self.company_id.id, False)),
+                '|', ('program_id.date_from', '=', False), ('program_id.date_from', '<=', fields.Date.context_today(self)),
                 '|', ('program_id.date_to', '=', False), ('program_id.date_to', '>=', fields.Date.context_today(self))]
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:397]:
```suggestion
        today = fields.Date.context_today(self):
        return [('active', '=', True),
                ('program_id.sale_ok', '=', True),
                ('company_id', 'in', (self.company_id.id, False)),
                '|', ('program_id.date_from', '=', False), ('program_id.date_from', '<=', today),
                '|', ('program_id.date_to', '=', False), ('program_id.date_to', '>=', today)]
```

File: addons/pos_loyalty/models/pos_config.py [hunk 24]
```diff
@@ -87,22 +87,26 @@ def use_coupon_code(self, code, creation_date, partner_id):
...
+        error_message = False
         if (coupon.expiration_date and coupon.expiration_date < check_date) or\
-            (coupon.program_id.date_to and coupon.program_id.date_to < fields.Date.context_today(self)) or\
+            (coupon.program_id.date_to and coupon.program_id.date_to < today_date) or\
             (coupon.program_id.limit_usage and coupon.program_id.total_order_count >= coupon.program_id.max_usage):
+            error_message = _('This coupon is expired (%s).', code)
+        elif coupon.program_id.date_from and coupon.program_id.date_from > today_date:
+            error_message = _('This coupon is not yet valid (%s).', code)
+        elif not coupon.program_id.reward_ids or\
+            not any(r.required_points <= coupon.points for r in coupon.program_id.reward_ids):
```
Comment (Feyensv) [inline addons/pos_loyalty/models/pos_config.py:99]:
```suggestion
        elif (
            not coupon.program_id.reward_ids
            or not any(r.required_points <= coupon.points for r in coupon.program_id.reward_ids)
        ):
```
Prefer the use of parentheses instead of line breaks ;)

File: addons/pos_loyalty/models/pos_config.py [hunk 25]
```diff
@@ -87,22 +87,26 @@ def use_coupon_code(self, code, creation_date, partner_id):
                 },
             }
         check_date = fields.Date.from_string(creation_date[:11])
+        today_date = fields.Date.context_today(self)
+        error_message = False
         if (coupon.expiration_date and coupon.expiration_date < check_date) or\
-            (coupon.program_id.date_to and coupon.program_id.date_to < fields.Date.context_today(self)) or\
+            (coupon.program_id.date_to and coupon.program_id.date_to < today_date) or\
             (coupon.program_id.limit_usage and coupon.program_id.total_order_count >= coupon.program_id.max_usage):
+            error_message = _('This coupon is expired (%s).', code)
```
Comment (Feyensv) [inline addons/pos_loyalty/models/pos_config.py:95]:
```suggestion
            error_message = _("This coupon is expired (%s).", code)
```
We follow the guideline of single-quoted strings for technical values, and double-quoted strings for strings shown to the user ;).

**Review by Feyensv — COMMENTED** [review]
Quick technical review, only minor linting comments for now (still gotta test the flow), good work :+1:

File: addons/sale_loyalty/models/sale_order.py [hunk 26]
```diff
@@ -989,10 +993,13 @@ def _try_apply_code(self, code):
                 return {'error': _('This coupon has already been used.')}
             program = coupon.program_id
 
+        today = fields.Date.context_today(self)
         if not program or not program.active:
             return {'error': _('This code is invalid (%s).', code), 'not_found': True}
+        elif (program.date_from and program.date_from > today):
+            return {'error': _('This code is not yet active (%s).', code)}
```
Comment (Demesmaeker) [inline addons/sale_loyalty/models/sale_order.py:1000]: I just realized, but... Why are these lines needed? You either pass through `_get_trigger_domain` for the coupon or through `_get_program_domain` for the program, both these searches includes the `date_from` and `date_to`.
How can you get to these lines? Aren't they dead code? :thinking:

Comment (Feyensv) [reply addons/sale_loyalty/models/sale_order.py]:
```python
    def _try_apply_code(self, code):
    
        base_domain = self._get_trigger_domain()
        domain = expression.AND([base_domain, [('mode', '=', 'with_code'), ('code', '=', code)]])
        rule = self.env['loyalty.rule'].search(domain)
        program = rule.program_id  # OK already filtered by the domain

        if not program:
            coupon = self.env['loyalty.card'].search([('code', '=', code)])
            program = coupon.program_id  # NOT OK, could be outside validity period
```

Not dead code :stuck_out_tongue:

**Review by Feyensv — COMMENTED** [review]
You should rebase to have the runbot doc & upgrade checks green ;)

File: addons/sale_loyalty/models/sale_order.py (hunk 26 above)
Comment (Demesmaeker) [reply addons/sale_loyalty/models/sale_order.py]: ...

```python

        if not program:
            coupon = self.env['loyalty.card'].search([('code', '=', code)])
            if not coupon.program_id.filtered_domain(self._get_program_domain()):
                return {'error': _('This code is invalid (%s).', code), 'not_found': True}
            program = coupon.program_id
```
Dead code? :thinking: (It's that time of the year again, I'm melting)

Comment (Feyensv) [reply addons/sale_loyalty/models/sale_order.py]: Ok, you win :ok_hand: 
(We're all (mad and) melting here)

**Review by Feyensv — APPROVED** [review]
@robodoo r+
